perception/sub8_perception/nodes/path_marker_finder.py

```diff
@@ -173,12 +198,62 @@ def cb_2d(self, req):
             res.found = True
         return res
 
+    def _update_transition_matrix(self, dt):
+        self.filter.transitionMatrix[0][4] = dt
+        self.filter.transitionMatrix[1][5] = dt
+        self.filter.transitionMatrix[2][6] = dt
+        self.filter.transitionMatrix[3][7] = dt
+
+    def _clear_filter(self, state):
+        self.found_count = 0
+        self.found = False
+        self.last3d = None
+        self.filter.errorCovPre = 1. * np.eye(self.state_size, dtype=np.float32)
+        if state is not None:
+            (x, y, z, theta) = state
+            state = np.array([[x], [y], [z], [theta], [0], [0], [0], [0]], dtype=np.float32)
+            self.filter.statePost = state
+
+    def _update_kf(self, (x, y, z, theta)):
+        if self.last_found_time is None:
+            self._clear_filter((x, y, z, theta))
+            return
+        dt = (self.image_sub.last_image_time - self.last_found_time).to_sec()
+        if dt <= 0 or dt > self.timeout_seconds:
+            rospy.logwarn("Timed out since last saw marker, resetting")
+            self._clear_filter((x, y, z, theta))
+            return
+        if self.last3d is not None:
+            if np.linalg.norm(np.array([x, y, z], dtype=np.float32) - self.last3d[0]) > 15:
+                self._clear_filter((x, y, z, theta))
+                rospy.logwarn("Too far apart, resetting")
+                return
+
+        self.found_count += 1
+        self._update_transition_matrix(dt)
+        measurement = 1.* np.array([x, y, z, theta], dtype=np.float32)
+        predict = self.filter.predict()
+        estimated = self.filter.correct(measurement)
```

Comment:
So reading the log from the pool day it sounds like you have issues since the whole thing isn't in your FOV at the same time. Maybe you could change from trying to track the whole marker to just trying to track the 4 corner points in the map frame. That would simplify the filtering down to just 4 points that don't move, and would make since with the problems you see. It could be harder to find the 4 corner points however.